### PR TITLE
updates internal framework for python functions 

### PR DESCRIPTION
**Pull request by ohasselblad**
@rochoa, could you take a look at this?

This branch updates the framework for accessing the Python functions in this extension. I stripped out the non-spatial kmeans function and will open a new feature branch for this function.

### CODE REVIEW

**Comment by ohasselblad** [comment]
@talos and @stuartlynn, could you give the file `src/py/crankshaft/crankshaft/clustering/kmeans.py` a code review here? There are some outstanding issues with writing unit tests, so I haven't added those to a commit yet (namely, it looks like I would have to add an additional class to mockplpy if I use the methods of the database response object like `db_response.colnames()`).

File: src/pg/sql/11_kmeans.sql [hunk 1]
```diff
@@ -1,21 +1,40 @@
...
+-- query: sql query to retrieve all the needed data
+
+CREATE OR REPLACE FUNCTION CDB_KMeansNonspatial(
+  query TEXT,
+  colnames TEXT[],
+  num_clusters INTEGER,
+  id_col TEXT DEFAULT 'cartodb_id',
+  standarize BOOLEAN DEFAULT true
+)
+RETURNS TABLE(cluster_label text, cluster_center text, rowid bigint) AS $$
```
Comment (stuartlynn) [inline src/pg/sql/11_kmeans.sql]: I am wondering if we should return the cluster center as a json object with keys for the variable name and its mean value in the cluster rather than text? Not sure how easy it is to return json from plpython

File: src/py/crankshaft/crankshaft/clustering/kmeans.py [hunk 2]
```diff
@@ -1,18 +1,98 @@
...
 
-    km = KMeans(n_clusters= no_clusters, n_init=no_init)
-    labels = km.fit_predict(zip(xs,ys))
-    return zip(ids,labels)
+    km = KMeans(n_clusters=no_clusters, n_init=no_init)
+    labels = km.fit_predict(zip(xs, ys))
+    return zip(ids, labels)
+
+
+def kmeans_nonspatial(query, colnames, num_clusters=5,
```
Comment (stuartlynn) [inline src/py/crankshaft/crankshaft/clustering/kmeans.py]: It might also be worthwhile calculating a [Silhouette](https://en.wikipedia.org/wiki/Silhouette_(clustering)) score for each cluster and returning that as well. This would give us a metric of how good each cluster is. You can do this relatively easily in sklearn : http://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_score.html

File: src/pg/sql/11_kmeans.sql (hunk 1 above)
Comment (talos) [inline src/pg/sql/11_kmeans.sql]: Just to but in -- even if JSON isn't properly supported, it should be straightforward to return stringified JSON from Plpython, then cast it back to JSON using a PL/PGSQL wrapper of the PL/Python function.

Comment (ohasselblad) [inline src/pg/sql/11_kmeans.sql]: I was thinking of stringifying it anyway since the data type mappings between sql and plpython are pretty limited, so will be doing what you suggest @talos

Comment (ohasselblad) [inline src/pg/sql/11_kmeans.sql]: Done, but the output in carto isn't optimal for display:
screen shot 2016-11-10 at 11 47 26

Outputting as a string will allow people to view the data in the builder, and my guess is that if people need to access it they may have the chops to know how to convert to a postgres json object and then parse it... On the other hand, we could ask for better display of json in the builder data view.

File: src/py/crankshaft/crankshaft/clustering/kmeans.py (hunk 2 above)
Comment (ohasselblad) [inline src/py/crankshaft/crankshaft/clustering/kmeans.py]: 👍 done :)

Comment (ohasselblad) [inline src/py/crankshaft/crankshaft/clustering/kmeans.py]: I actually did `metrics.silhouette_sample` http://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_samples.html

File: src/py/crankshaft/crankshaft/query_runner.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
+"""class for fetching data"""
+import plpy
+
+
+class QueryRunner:
```
Comment (rochoa) [inline src/py/crankshaft/crankshaft/query_runner.py]: As this is attached to moran/markov queries, I would go with a name closer to that. QueryRunner is quite generic, while get_markov is very specific. What about something like `AnalysisDataProvider`?

File: src/py/crankshaft/crankshaft/query_runner.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    def get_markov(self, query):
```
Comment (rochoa) [inline src/py/crankshaft/crankshaft/query_runner.py]: Why do you need to provide the query? I would try to encapsulate the query generation within this class, making this method to handle the required params (num_ngbrs,...).

File: src/py/crankshaft/crankshaft/query_runner.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            data = plpy.execute(query)
+
+            if len(data) == 0:
+                return pu.empty_zipped_array(4)
+
+            return data
+        except plpy.SPIError, err:
+            plpy.error('Analysis failed: %s' % err)
+
+    def get_moran(self, query):
```
Comment (rochoa) [inline src/py/crankshaft/crankshaft/query_runner.py]: Similar approach in all the public functions in this class.

File: src/py/crankshaft/crankshaft/query_runner.py (hunk 3 above)
Comment (ohasselblad) [inline src/py/crankshaft/crankshaft/query_runner.py]: 👍 I like `AnalysisDataProvider`

File: src/py/crankshaft/crankshaft/query_runner.py (hunk 4 above)
Comment (ohasselblad) [inline src/py/crankshaft/crankshaft/query_runner.py]: Ok, makes sense. And `__init__` would construct the query.. Which would clean up the code in other functions like `Moran.local_stat` :)

File: src/py/crankshaft/crankshaft/query_runner.py (hunk 5 above)
Comment (ohasselblad) [inline src/py/crankshaft/crankshaft/query_runner.py]: done.

File: src/py/crankshaft/crankshaft/query_runner.py (hunk 4 above)
Comment (ohasselblad) [inline src/py/crankshaft/crankshaft/query_runner.py]: Done

**Comment by ohasselblad** [comment]
Ok @rochoa, made those changes. What do you think?

**Comment by rochoa** [comment]
Sorry for the delay @ohasselblad.

It looks better for me 👍 .

**Comment by ohasselblad** [comment]
@ethervoid I updated one function from a new merge that didn't have the new structure yet, and removed `plpy` imports that are no longer needed. Looks like there's another failure for the postgres dependency (`postgresql-9.5=9.5.2-3cdb2`). Otherwise, looks good!
